aiohttp/client.py

```diff
@@ -411,37 +394,30 @@ def request(self, method, url, **kwargs):
                 resp.raise_for_status()
 
             # register connection
-            if handle is not None:
-                if resp.connection is not None:
-                    resp.connection.add_callback(handle.cancel)
-                else:
-                    handle.cancel()
+            #  XXX this will only be required if there are still valid open timeouts after request_end
+            if resp.connection is not None:
+                resp.connection.add_callback(lifecycle.clear_timeouts)
+            else:
+                lifecycle.clear_timeouts()
 
             resp._history = tuple(history)
 
-            for trace in traces:
-                await trace.send_request_end(
-                    method,
-                    url,
-                    headers,
-                    resp
-                )
+            await lifecycle.send_request_end(
+                method,
+                url,
+                headers,
+                resp
+            )
             return resp
 
         except BaseException as e:
-            # cleanup timer
-            tm.close()
-            if handle:
-                handle.cancel()
-                handle = None
-
-            for trace in traces:
-                await trace.send_request_exception(
-                    method,
-                    url,
-                    headers,
-                    e
-                )
+            await lifecycle.send_request_exception(
+                method,
+                url,
+                headers,
+                e
+            )
+            lifecycle.clear_timeouts()
```

Comment:
can we do this as part of a context, it's too easy to get wrong otherwise

Reply:
I'm not sure whether it is required at all, `clear_timeouts` is analogous to the old `handle.cancel` e.g. also used in (resp.connection.add_callback(handle.cancel))[https://github.com/aio-libs/aiohttp/pull/2842/files#diff-7dd84b5ef8d5eea2de1dfc5329411dfcR397]. As timeouts are automatically cancelled once the end signal is emitted, we probably don't need any clean-up any more. But I'm not 100% sure whether there were any side effects through `resp.connection.add_callback` that I'm not aware of right now.

Reply:
cool, ya lets figure that out